integrations/tensorflow/e2e/math/math_test.py

```diff
@@ -0,0 +1,570 @@
+import collections
+import os
+from typing import Any, Dict, Sequence, Type, Union
+
+from absl import app
+from absl import flags
+import numpy as np
+from pyiree.tf.support import tf_test_utils
+from pyiree.tf.support import tf_utils
+import tensorflow.compat.v2 as tf
+
+FLAGS = flags.FLAGS
+
+
+class Config:
+  """Specifies a unittest."""
+
+  def __init__(self,
+               signature: Sequence[tf.TensorSpec],
+               args: Sequence[Any] = None,
+               kwargs: Dict[str, Any] = None,
+               generator: tf_utils.InputGeneratorType = None):
+    self.signature = signature
+    self.args = args
+    self.kwargs = kwargs if kwargs is not None else dict()
+    self.generator = generator
+
+
+class ArgsConfig(Config):
+  """Uses input arguments to generate an input signature."""
+
+  def __init__(self,
+               args: Sequence[Any],
+               kwargs: Dict[str, Any] = None,
+               generator: tf_utils.InputGeneratorType = None):
+    signature = tf_utils.apply_function(
+        args, lambda x: tf.TensorSpec.from_tensor(tf.convert_to_tensor(x)))
+    super().__init__(signature, args, kwargs, generator)
+
+
+class DtypeConfig(Config):
+  """Specifies a unittest for each dtype in dtypes."""
+
+  def __init__(self,
+               signature: Sequence[tf.TensorSpec],
+               args: Sequence[Any] = None,
+               kwargs: Dict[str, Any] = None,
+               generator: tf_utils.InputGeneratorType = None,
+               dtypes: Sequence[tf.DType] = None):
+    super().__init__(signature, args, kwargs, generator)
+    for dtype in dtypes:
+      if not isinstance(dtype, tf.DType):
+        raise TypeError(f"All dtypes must be tf.DTypes, but got '{dtype}'.")
+    self.dtypes = dtypes
+
+  def __iter__(self):
+    for dtype in self.dtypes:
+      signature = tf_utils.apply_function(
+          self.signature, lambda spec: tf.TensorSpec(spec.shape, dtype))
+      dtype_config = Config(signature, self.args, self.kwargs, self.generator)
+      yield dtype.name, dtype_config
```

Comment:
Seems like these could be factory functions instead of subclasses? Especially when you are doing isinstance queries below it becomes pretty clear that DtypeConfig isn't really a Config at all (see https://en.wikipedia.org/wiki/Liskov_substitution_principle).


```suggestion
class Config:
  """Specifies a unittest."""

  def __init__(self,
               signature: Sequence[tf.TensorSpec],
               args: Sequence[Any] = None,
               kwargs: Dict[str, Any] = None,
               generator: tf_utils.InputGeneratorType = None):
    self.signature = signature
    self.args = args
    self.kwargs = kwargs if kwargs is not None else dict()
    self.generator = generator
    
  @classmethod
  def FromArgs(args: Sequence[Any],
               kwargs: Dict[str, Any] = None,
               generator: tf_utils.InputGeneratorType = None):
    signature = tf_utils.apply_function(
        args, lambda x: tf.TensorSpec.from_tensor(tf.convert_to_tensor(x)))
    return Config(signature, args, kwargs, generator)
    
  @classmethod
  def ForDtypes(signature: Sequence[tf.TensorSpec],
                args: Sequence[Any] = None,
                kwargs: Dict[str, Any] = None,
                generator: tf_utils.InputGeneratorType = None,
                dtypes: Sequence[tf.DType] = None):
      configs = {}
      for dtype in self.dtypes:
        signature = tf_utils.apply_function(
            self.signature, lambda spec: tf.TensorSpec(spec.shape, dtype))
        dtype_config = Config(signature, self.args, self.kwargs, self.generator)
        configs[dtype.name] = dtype_config
```